hoomd/mpcd/RejectionVirtualParticleFillerGPU.cuh

```diff
@@ -0,0 +1,265 @@
+ * \param d_track_bounded_particles Particle tracking - in/out of given geometry
+ * \param lo Left extrema of the sim-box
+ * \param hi Right extrema of the sim-box
+ * \param first_tag First tag (rng argument)
+ * \param vel_factor Scale factor for uniform normal velocities consistent with particle mass / temperature
+ * \param filler_id Identifier for the filler (rng argument)
+ * \param type Particle type for filling
+ * \param N_virt_max Maximum no. of virtual particles that can exist
+ * \param timestep Current timestep
+ * \param seed User seed for RNG
+ *
+ * \tparam Geometry type of the confined geometry \a geom
+ *
+ * \b implementation
+ * We assign one thread per particle to draw random particle positions within the box and velocities consistent
+ * with system temperature. Along with this a boolean array tracks if the particles are in/out of bounds of the
+ * given geometry.
+ */
+template<class Geometry>
+__global__ void draw_virtual_particles(Scalar4 *d_tmp_pos,
+                                       Scalar4 *d_tmp_vel,
+                                       bool *d_track_bounded_particles,
+                                       const Scalar3 lo,
+                                       const Scalar3 hi,
+                                       const unsigned int first_tag,
+                                       const Scalar vel_factor,
+                                       const unsigned int filler_id,
+                                       const unsigned int type,
+                                       const unsigned int N_virt_max,
+                                       const unsigned int timestep,
+                                       const unsigned int seed,
+                                       const unsigned int block_size,
+                                       const Geometry geom)
+    {
+    // one thread per particle
+    const unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx >= N_virt_max)
+        return;
+
+    // initialize random number generator for positions and velocity
+    hoomd::RandomGenerator rng(hoomd::RNGIdentifier::RejectionFiller, seed, timestep, first_tag+idx, filler_id);
+    d_tmp_pos[idx] = make_scalar4(hoomd::UniformDistribution<Scalar>(lo.x, hi.x)(rng),
+                                  hoomd::UniformDistribution<Scalar>(lo.y, hi.y)(rng),
+                                  hoomd::UniformDistribution<Scalar>(lo.z, hi.z)(rng),
+                                  __int_as_scalar(type));
+
+    hoomd::NormalDistribution<Scalar> gen(vel_factor, 0.0);
+    Scalar3 vel;
+    gen(vel.x, vel.y, rng);
+    vel.z = gen(rng);
+    d_tmp_vel[idx] = make_scalar4(vel.x,
+                                  vel.y,
+                                  vel.z,
+                                  __int_as_scalar(mpcd::detail::NO_CELL));
+
+    // check if particle is inside/outside the confining geometry
+    d_track_bounded_particles[idx] = geom.isOutside(make_scalar3(d_tmp_pos[idx].x,
```

Comment:
I think it would be a good idea to explicitly generate a `Scalar3 pos` variable before line 136, then use that to write to `d_tmp_pos` and as the argument to `isOutside`. You should also move this code up before the velocity generation so that the compiler is clearly allowed to free the registers being used for the positions. The compiler may already do a lot of this, though.